File: backend/app/services/notification_service.py
from __future__ import annotations

from sqlalchemy.ext.asyncio import AsyncSession

from app.models.notification import Notification


async def create_notification(
    db: AsyncSession, *, user_id: str, type: str, title: str, message: str
) -> Notification:
    notif = Notification(user_id=user_id, type=type, title=title, message=message)
    db.add(notif)
    await db.commit()
    await db.refresh(notif)
    return notif


async def notify_breach(db: AsyncSession, *, user_id: str, company_name: str, severity: str) -> Notification:
    return await create_notification(
        db,
        user_id=user_id,
        type="breach",
        title=f"New breach detected: {company_name}",
        message=f"{company_name} was found in a {severity}-severity data breach. Review your data ledger entry.",
    )


async def notify_deadline(db: AsyncSession, *, user_id: str, company_name: str, days_left: int) -> Notification:
    return await create_notification(
        db,
        user_id=user_id,
        type="deadline",
        title=f"Response deadline approaching: {company_name}",
        message=f"{company_name} has {days_left} day(s) left to respond to your deletion request.",
    )


async def notify_deletion_status(
    db: AsyncSession, *, user_id: str, company_name: str, status: str
) -> Notification:
    type_ = "deletion_approved" if status == "resolved" else "deletion_failed" if status == "escalated" else "info"
    title = f"Deletion request update: {company_name}"
    message = f"Your deletion request to {company_name} is now '{status}'."
    return await create_notification(db, user_id=user_id, type=type_, title=title, message=message)

File: backend/src/main/java/com/infohub/backend/service/NotificationService.java
package com.infohub.backend.service;

import com.infohub.backend.entity.Notification;
import com.infohub.backend.entity.User;
import com.infohub.backend.repository.NotificationRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;

@Service
public class NotificationService {

    private final NotificationRepository notificationRepository;

    @Autowired
    public NotificationService(NotificationRepository notificationRepository) {
        this.notificationRepository = notificationRepository;
    }

    public List<Notification> getNotifications(Long userId) {
        return notificationRepository.findByUserIdOrderByCreatedAtDesc(userId);
    }

    public long getUnreadCount(Long userId) {
        return notificationRepository.countByUserIdAndIsReadFalse(userId);
    }

    public void markRead(Long id, Long userId) {
        notificationRepository.findById(id).ifPresent(n -> {
            if (n.getUser().getId().equals(userId)) {
                n.setRead(true);
                notificationRepository.save(n);
            }
        });
    }

    @Transactional
    public void markAllRead(Long userId) {
        notificationRepository.markAllReadByUserId(userId);
    }

    public void createBreachAlert(User user, String companyName) {
        Notification n = Notification.builder()
                .user(user)
                .type(Notification.NotificationType.BreachAlert)
                .title("Data Breach Detected: " + companyName)
                .body("Your data was found in a breach at " + companyName +
                      ". Change your password and enable 2FA immediately.")
                .actionUrl("/breach-monitor")
                .build();
        notificationRepository.save(n);
    }

    public void createDeadlineReminder(User user, String companyName, int daysLeft) {
        Notification n = Notification.builder()
                .user(user)
                .type(Notification.NotificationType.DeadlineReminder)
                .title("Deletion Deadline: " + daysLeft + " day(s) left — " + companyName)
                .body(companyName + " has " + daysLeft + " day(s) to respond to your deletion request.")
                .actionUrl("/deletion-requests")
                .build();
        notificationRepository.save(n);
    }

    public void createStatusUpdate(User user, String companyName, String status) {
        Notification n = Notification.builder()
                .user(user)
                .type(Notification.NotificationType.StatusUpdate)
                .title("Deletion Request Updated: " + companyName)
                .body("Your deletion request for " + companyName + " has been updated to: " + status)
                .actionUrl("/deletion-requests")
                .build();
        notificationRepository.save(n);
    }

    public void createAIRecommendation(User user, String message) {
        Notification n = Notification.builder()
                .user(user)
                .type(Notification.NotificationType.AIRecommendation)
                .title("Privacy Recommendation")
                .body(message)
                .actionUrl("/risk-analysis")
                .build();
        notificationRepository.save(n);
    }
}
